src/ts/ConvivaAnalyticsTracker.ts

```diff
@@ -0,0 +1,939 @@
+   * production or automated testing.
+   */
+  gatewayUrl?: string;
+
+  /**
+   * Option to set the Conviva Device Category, which is used to assist with
+   * user agent string parsing by the Conviva SDK. (default: WEB)
+   * @deprecated Use `deviceMetadata.category` field
+   */
+  deviceCategory?: Conviva.valueof<Conviva.ConvivaConstants['DeviceCategory']>;
+
+  /**
+   * Option to override the Conviva Device Metadata.
+   * (Default: Auto extract all options from User Agent string)
+   */
+  deviceMetadata?: {
+    /**
+     * Option to set the Conviva Device Category, which is used to assist with
+     * user agent string parsing by the Conviva SDK.
+     * (default: The same specified in config.deviceCategory)
+     */
+    category?: Conviva.valueof<Conviva.ConvivaConstants['DeviceCategory']>;
+
+    /**
+     * Option to override the Conviva Device Brand.
+     * (Default: Auto extract from User Agent string)
+     */
+    brand?: string;
+
+    /**
+     * Option to override the Conviva Device Manufacturer.
+     * (Default: Auto extract from User Agent string)
+     */
+    manufacturer?: string;
+
+    /**
+     * Option to override the Conviva Device Model.
+     * (Default: Auto extract from User Agent string)
+     */
+    model?: string;
+
+    /**
+     * Option to override the Conviva Device Type
+     * (Default: Auto extract from User Agent string)
+     */
+    type?: Conviva.valueof<Conviva.ConvivaConstants['DeviceType']>;
+
+    /**
+     * Option to override the Conviva Device Version.
+     * (Default: Auto extract from User Agent string)
+     */
+    version?: string;
+
+    /**
+     * Option to override the Conviva Operating System Name
+     * (Default: Auto extract from User Agent string)
+     */
+    osName?: string;
+
+    /**
+     * Option to override the Conviva Operating System Version
+     * (Default: Auto extract from User Agent string)
+     */
+    osVersion?: string;
+  };
+}
+
+export interface EventAttributes {
+  [key: string]: string;
+}
+
+export class ConvivaAnalyticsTracker {
+  private static readonly VERSION: string = '{{VERSION}}';
+
+  private static readonly STALL_TRACKING_DELAY_MS = 100;
+  private readonly player: PlayerAPI;
+  private readonly events: typeof PlayerEvent;
+  private readonly handlers: PlayerEventWrapper;
+  private readonly config: ConvivaAnalyticsConfiguration;
+  private readonly contentMetadataBuilder: ContentMetadataBuilder;
+
+  private readonly logger: Conviva.LoggingInterface;
+  private sessionKey: number;
+  private convivaVideoAnalytics: Conviva.VideoAnalytics;
+  private convivaAdAnalytics: Conviva.AdAnalytics;
+
+  /**
+   * Tracks the ad break status and is true between ON_AD_STARTED and ON_AD_FINISHED/SKIPPED/ERROR.
+   * This flag is required because player.isAd() is unreliable and not always true between the events.
+   */
+  private _isAdBreakActive: boolean;
```

Comment:
What about using [player.ads.getActiveAdBreak()](https://cdn.bitmovin.com/player/web/8/docs/interfaces/Advertising.PlayerAdvertisingAPI.html#getActiveAdBreak) instead, is that reliable? Or would that not work now anyway due to SSAI?

Reply:
`isAdBreak` was in the previous implementation, I simply followed it, so not sure if [player.ads.getActiveAdBreak()](https://cdn.bitmovin.com/player/web/8/docs/interfaces/Advertising.PlayerAdvertisingAPI.html#getActiveAdBreak) is reliable. Anyways, we need this flag because it also tacks SSAI, so we cannot use `getActiveAdBreak`.

Reply:
Makes sense regarding SSAI 👍